### PR TITLE
pilot: refactor version to avoid globals

### PR DESCRIPTION
**Pull request by howardjohn**
Minor refactoring, no need to use globals here. Also clean up to use
atomics instead of mutex for simplicity

### CODE REVIEW

**Review by zirain — APPROVED** [review]

File: pilot/pkg/xds/discovery.go [hunk 1]
```diff
@@ -657,3 +643,22 @@ func (s *DiscoveryServer) WaitForRequestLimit(ctx context.Context) error {
...
+	// versionNum counts versions
+	versionNum atomic.Uint64
+}
+
+func (v PushVersion) NextVersion() string {
+	return time.Now().Format(time.RFC3339) + "/" + strconv.FormatUint(v.versionNum.Inc(), 10)
+}
+
+func (v PushVersion) SetVersion(s string) {
+	v.version.Store(s)
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/discovery.go:659]: Can you add a test, i doubt this is not right as v is not a pointer receiver

File: pilot/pkg/xds/ads.go [hunk 2]
```diff
@@ -833,7 +833,7 @@ func (s *DiscoveryServer) ProxyUpdate(clusterID cluster.ID, ip string) {
 // AdsPushAll will send updates to all nodes, with a full push.
 // Mainly used in Debug interface.
 func AdsPushAll(s *DiscoveryServer) {
-	s.AdsPushAll(versionInfo(), &model.PushRequest{
+	s.AdsPushAll(s.PushVersion.CurrentVersion(), &model.PushRequest{
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:836]: We can get the pushversion from pushContext

File: pilot/pkg/xds/discovery.go [hunk 3]
```diff
@@ -283,7 +277,7 @@ func (s *DiscoveryServer) Push(req *model.PushRequest) {
 	if !req.Full {
 		req.Push = s.globalPushContext()
 		s.dropCacheForRequest(req)
-		s.AdsPushAll(versionInfo(), req)
+		s.AdsPushAll(s.PushVersion.CurrentVersion(), req)
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/discovery.go:280]: same here we can get version from push context

File: pilot/pkg/xds/ads.go (hunk 2 above)
Comment (howardjohn) [reply pilot/pkg/xds/ads.go]: Good idea. I fixed the bug and simplified it a lot with your suggestion

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**